### DISCUSSION TITLE
Separating 2 shots that use the same switch?

### DISCUSSION THREAD

**Discussion by ironspider** [missionpinball/missionpinball#35] [General]
Greetings all, so my machine has a ramp (a Baywatch dual right ramp) that can be hit by either a shot from the left upper flipper, or the lower flippers. The ramp has 2 switches on it: "s_right_ramp_enter" (which is down at the bottom of the right lower entry to the ramp) and "s_right_ramp_exit" (which is at the back of the ramp curve before it hits the wireform to return to the left lower flipper). Here is my dilemma...

During MB, the super jackpot should be a shot from the left upper flipper to the ramp. This will only hit the s_right_ramp_exit switch and be returned to the lower left flipper, sending the state machine back to normal MB mode (like AFM). But when the super jackpot is lit, a shot to the lower right part of the ramp (doable from either lower flipper) *also* hits that "s_right_ramp_exit" switch. I don't want shots from the lower flippers to award the super jackpot so how do I do this so that the machine can differentiate between a shot from the lower flippers hitting s_right_ramp_enter and s_right_ramp_exit and the upper left flipper which is only hitting "s_right_ramp_exit"? 

Is there way to say something like "super jackpot shot is s_right_ramp exit and NOT s_ramp_ramp_enter within a second earlier"? Something to that effect? Or something like when that super jackpot mode is running a shot to s_right_ramp_enter disables the s_right_ramp_exit for some amount of time (like half a second)?

Thoughts?

**Comment by Pucksnok** [comment]
Not sure if there are any function ready in mpf, it may be. But if there weren't any I would in my multiball mode start a timer when s_right_ramp_enter is activated, for example 1 second. The exit switch would then stop the timer, if the timer is completed after 1 second, award the jackpot.

**Reply by ironspider** [reply]
I'm trying to wrap my head around this but I'm confused. I want only a shot from the upper left flipper to score the Super Jackpot by tripping the "s_right_ramp_exit" switch. If I set a timer that is triggered when the enter switch (s_right_ramp_enter) is triggered then how does that help me? Because won't I need to *start* the timer (which requires a shot from the lower flippers) to do anything with it?

Basically my setup is that when MB starts, it is a state machine. When you hit the 4 regular jackpot shots the state machine calls the super jackpot mode. I only want a shot from the upper left flipper to score the jackpot (and the only switch the upper left flipper can hit is s_right_ramp_exit). Both of the lower flippers, however, can make the right ramp which hits both s_right_ramp_enter and s_right_ramp_exit. So s_right_ramp_exit is being hit both ways.

How do I write something that, in effect, says "ignore a hit to s_right_ramp_exit if s_right_ramp_enter was hit right before it"

**Reply by Pucksnok** [reply]
Ah, sorry. I read it the other way around.

If it's the exit switch that both upper and lower flippers hit I would try and do it like this.

1. If entry switch is hit, change a variable to 1.
2. Start a timer that resets above variable. 
3. Also reset variable when exit switch is hit (if it's a multiball you may hit the exit switch almost simultaneously so you want the jackpot to trigger once).
4. Make a condition and make the jackpot trigger only if variable is 0.

May be some other ways but this is how I would do it. You may also think how to do it if two balls enter at entry switch almost simultaneously. Probably I would let the variable go up to 2 and 3 and let the timer reset it completely. I have a hangover so can't really get my English and thoughts right, sorry. 😂

**Reply by ironspider** [reply]
Thank you so much for this reply! After you initial reply I did eventually come to the conclusion "I think they think I want the opposite" but that did start the gears turning about how to do it. I too was thinking about a variable that gets modified just like you posted! So that's a good sanity check for me. I'll try it out and let you know how it goes!

**Reply by ironspider** [reply]
Okay I now think I have it working. I have a MB mode that uses a state machine and when all 4 normal MB jackpots have been hit the state machine shifts to super jackpot state by calling the super jackpot mode. Then, using your ideas I finally crafted a solution that:
- Sets a variable "super_jackpot_ready" to 0
- Any time the right ramp enter switch is hit it starts (I actually user reset, not start) a 1second timer and adds 1 to the super_jackpot_ready variable.
- When the timer completes OR the right ramp exit switch is hit it sets the super_jackpot_ready variable to 0 using a variable_player entry
- Then, in the shot definition for the super jackpot shot, I removed the "switch: s_right_ramp_exit_active" entry and replaced it with a "hit_events:" entry that says "s_right_ramp_exit_active{current_player.super_jackpot_ready==0}"

I've tested it 6-7 times and it seems to be working exactly how I want it to! Thanks!
